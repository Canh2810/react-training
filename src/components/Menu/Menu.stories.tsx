import { StoryObj, Meta } from '@storybook/react'
import { MemoryRouter } from 'react-router-dom'
import Menu from '.'
import { MenuVariants } from '@types'
import { LIST_MENU_ITEM } from '@constants'

export default {
  title: 'Components/Menu',
  component: Menu,
  decorators: [
    (Story) => (
      <MemoryRouter>
        <Story />
      </MemoryRouter>
    ),
  ],
} as Meta

type Story = StoryObj<typeof Menu>

export const Default: Story = {
  args: {
    listItems: LIST_MENU_ITEM,
  },
}

export const Vertical: Story = {
  args: {
    listItems: LIST_MENU_ITEM,
    variant: MenuVariants.Vertical,
  },
}
