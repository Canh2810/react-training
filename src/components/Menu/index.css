.menu__item a {
  text-decoration: none;
  font-size: var(--sm-font);
  font-weight: var(--normal);
  color: var(--text-primary-color);
  font-size: var(--tertiary-font);
  line-height: 24px;
}

.menu__item:hover {
  opacity: 0.7;
}

.menu__item--horizontal {
  font-size: var(--primary-font);
}

.menu__item--horizontal--dark {
  color: var(--text-secondary-color);
}
