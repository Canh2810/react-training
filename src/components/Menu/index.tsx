// Libs
import { CSSProperties, memo } from 'react'
import { Link } from 'react-router-dom'

// Types
import { IMenuItem, MenuVariants } from '@types'

// Stores
import { useThemeStore } from '@stores'

// Themes
import { variants } from '@themes'

// Style
import './index.css'

export interface MenuProps {
  listItems: IMenuItem[]
  variant?: MenuVariants
  className?: string
}

const Menu = ({
  listItems,
  variant = MenuVariants.Horizontal,
  className = '',
}: MenuProps) => {
  const { theme } = useThemeStore((state) => state)

  const menuStyle = {
    ...(variants[variant] as CSSProperties),
  }

  return (
    <ul className={`menu ${className}`} style={{ ...menuStyle }}>
      {listItems.map(({ title, href }: IMenuItem) => (
        <li
          key={title}
          className={`menu__item menu__item--${variant} menu__item--${variant}--${theme}`}
        >
          <Link to={href}>{title}</Link>
        </li>
      ))}
    </ul>
  )
}

export default memo(Menu)
