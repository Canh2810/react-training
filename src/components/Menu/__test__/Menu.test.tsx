import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import Menu from '..'
import { MenuVariants } from '@types'
import { LIST_MENU_ITEM } from '@constants'

describe('Menu component', () => {
  it('should render correctly with variant is horizontal', () => {
    const { container } = render(
      <MemoryRouter>
        <Menu listItems={LIST_MENU_ITEM} />
      </MemoryRouter>,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render correctly with variant is vertical', () => {
    const { container } = render(
      <MemoryRouter>
        <Menu listItems={LIST_MENU_ITEM} variant={MenuVariants.Vertical} />
      </MemoryRouter>,
    )
    expect(container).toMatchSnapshot()
  })

  it('calls the correct onClick handler when a menu item is clicked', () => {
    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/about"
            element={<div data-testid="about-page">About Page</div>}
          />
        </Routes>
        <Menu listItems={LIST_MENU_ITEM} />
      </MemoryRouter>,
    )

    const menuItem = screen.getByText('About')

    fireEvent.click(menuItem)
    const aboutPage = screen.getByTestId('about-page')
    expect(aboutPage).toBeInTheDocument()
  })
})
