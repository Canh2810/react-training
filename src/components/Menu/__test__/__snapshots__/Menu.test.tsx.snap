// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component should render correctly with variant is horizontal 1`] = `
<div>
  <ul
    class="menu "
    style="display: flex; align-items: center; justify-content: center; gap: 30px;"
  >
    <li
      class="menu__item menu__item--horizontal menu__item--horizontal--light"
    >
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li
      class="menu__item menu__item--horizontal menu__item--horizontal--light"
    >
      <a
        href="/about"
      >
        About
      </a>
    </li>
    <li
      class="menu__item menu__item--horizontal menu__item--horizontal--light"
    >
      <a
        href="/contact"
      >
        Contact
      </a>
    </li>
  </ul>
</div>
`;

exports[`Menu component should render correctly with variant is vertical 1`] = `
<div>
  <ul
    class="menu "
    style="display: flex; align-items: center; justify-content: center; gap: 8px; flex-direction: column;"
  >
    <li
      class="menu__item menu__item--vertical menu__item--vertical--light"
    >
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li
      class="menu__item menu__item--vertical menu__item--vertical--light"
    >
      <a
        href="/about"
      >
        About
      </a>
    </li>
    <li
      class="menu__item menu__item--vertical menu__item--vertical--light"
    >
      <a
        href="/contact"
      >
        Contact
      </a>
    </li>
  </ul>
</div>
`;
